### DISCUSSION TITLE
Installing Compiling & Programming problems

### DISCUSSION THREAD

**Discussion by vitalyxxl** [dzurikmiroslav/esp32-evse#8] [Q&A]
Please help me with compile project .

**Comment by dzurikmiroslav** [comment]
Default ESP-IDF configuration are set to not use UART0 as standard log ouput. All uarts ports are managed by firmware for user configuration (AT commands, Modbus, System logs...).
Beware, when you configure serial as System logs, at startup not all logs are printed until serial are initialized.
To set UART0 as standard output need do following things.

In `sdkoconfig.defaults` are written:
```
CONFIG_ESP_CONSOLE_NONE=y
CONFIG_ESP_CONSOLE_UART_NUM=-1
```
So you need edit `sdkconfig` to (or in SDK Configuration editor set **Channel for console output** to **Default**):
```
CONFIG_ESP_CONSOLE_UART_DEFAULT=y
# CONFIG_ESP_CONSOLE_UART_CUSTOM is not set
# CONFIG_ESP_CONSOLE_NONE is not set
CONFIG_ESP_CONSOLE_UART=y
CONFIG_ESP_CONSOLE_MULTIPLE_UART=y
CONFIG_ESP_CONSOLE_UART_NUM=0
CONFIG_ESP_CONSOLE_UART_BAUDRATE=115200
```
In 'main.c' comment row `esp_log_set_vprintf(log_vprintf);` and in board config omit `SERIAL_X` definition.

**Comment by fototakas** [comment]
I can confirm compilation is working OK, just follow espressif IDF setup guide for your OS:
https://docs.espressif.com/projects/esp-idf/en/v5.0/esp32/get-started/windows-setup.html#
Go thru the all guide, try to compile hello_world example. After success with hello_world, try the same steps with esp32-evse project.  Do not forget to do https://github.com/dzurikmiroslav/esp32-evse/wiki/Board-config  ("idf.py menuconfig" in IDF setup guide) And you should get working flash file.
After flashing and ESP32 reboot, shortly take down (connect to the ground) GPIO-0 pin  ( default board.cfg
BUTTION_WIFI_GPIO pin) on your board and you should get WIFI access point with name esve-XXXXXX, after connection, go to ip 192.168.4.1 there should see WebUI.

**Comment by vitalyxxl** [comment]
All these changes and fixes have been made by . Rebuilt the project. The port monitor now displays debugging information about the program's operation.
SPIFFS mounting error is generated when starting the internal program
I tried to flash on several options for debug boards. The error is the same.
What do you say, gentlemen?
I am attaching a screenshot with the log of the board. The computer tries to reboot cyclically and gives the same sequence of messages (in the port monitor). I presented one of these cycles.
![ERROR_EVSE_ESP32](https://user-images.githubusercontent.com/37896761/215272107-d8981aef-4fff-4b8a-a440-b580bcd4829c.jpg)

**Reply by fototakas** [reply]
This could be a hardware issue (flash), do I understand you correctly, you tried different ESP32 boards and result is the same?
Was those chips completely new or already used for some test\projects?
Try to erase flash before uploading evse firmaware:
esptool.py --port "you com port" erase_flash
I tried on 2 new boards, had another issue with adc calibration, but no problems with SPIFFS mount.

**Reply by vitalyxxl** [reply]
All boards have been tested on other projects. Starting with examples that come with the ESP-IDF package and ending with complex projects using SPIFFS. Everything works as it should.
The clear flash command didn't help. the application still generates the error.

PS: in Arduino, all projects are also remarkably compiled  and work according to the laid down algorithm.

I guess I need to get some rest.....

**Reply by fototakas** [reply]
Please try to flash my compiled files:
https://www106.zippyshare.com/v/bwSMRsUs/file.html
Download/extract zip and change /dev/ttyUSB0 to your com port, then flash:
`esptool esp32 -p /dev/ttyUSB0 -b 460800 --before=default_reset --after=hard_reset write_flash --flash_mode dio --flash_freq 40m --flash_size 4MB 0x1000 bootloader.bin 0x10000 esp32-evse.bin 0x8000 partition-table.bin 0xd000 ota_data_initial.bin 0x370000 cfg.bin`

**Reply by vitalyxxl** [reply]
Flashing by "ESP32 FLASH DOWNLOAD TOOL V3.9.3"  - OK 
ESP32DEV KIT restarting ....  
connecting  (WIFI AP) 
192.168.4.1 OK 
UI started & working OK 

RESULT OF THIS TEST :
   My ESP32 boards is OK

**Reply by fototakas** [reply]
Yes, your board is OK, bud something wrong with compilation process.  Miroslav and me are compiling on Linux, you use Windows.
Maybe you can try to use free virtualization software on windows like https://www.virtualbox.org/ and install Ubuntu Linux for this. It is quite easy process.

**Reply by vitalyxxl** [reply]
Thank you for the sent package of .bin files!
Using a special utility for ESP32 firmware and the resulting files, we managed to write all the data to the board. The access point appears and everything works as expected.

Experimentally, it turned out that when building the project, for some reason, not all .bin files, but only "bootloader.bin" and "esp32-evse.bin".

**Reply by dzurikmiroslav** [reply]
Good works 👍 This looks like a bug on Windows... But unfortunately I don't have Win OS. To solve this issue I must create virtual Win environment, this take me more time... Currently I am refactoring error handling...

**Reply by dzurikmiroslav** [reply]
I had to install windows 10, from VSCode install ESP extension. Project was builded successfully and after flashing I had no SPIFFS error :-/ Startup log:
```
entry 0x4008064c
I (27) boot: ESP-IDF v5.0-494-g490216a2ac 2nd stage bootloader
I (27) boot: compile time 22:09:15
I (27) boot: chip revision: v3.0
I (32) boot_comm: chip revision: 3, min. bootloader chip revision: 0
I (39) boot.esp32: SPI Speed      : 40MHz
I (43) boot.esp32: SPI Mode       : DIO
I (48) boot.esp32: SPI Flash Size : 4MB
I (52) boot: Enabling RNG early entropy source...
I (58) boot: Partition Table:
I (61) boot: ## Label            Usage          Type ST Offset   Length
I (69) boot:  0 nvs              WiFi data        01 02 00009000 00004000
I (76) boot:  1 otadata          OTA data         01 00 0000d000 00002000
I (84) boot:  2 phy_init         RF data          01 01 0000f000 00001000
I (91) boot:  3 factory          factory app      00 00 00010000 00120000
I (98) boot:  4 ota_0            OTA app          00 10 00130000 00120000
I (106) boot:  5 ota_1            OTA app          00 11 00250000 00120000
I (113) boot:  6 cfg              Unknown data     01 82 00370000 00019000
I (121) boot: End of partition table
I (125) boot: Defaulting to factory image
I (130) boot_comm: chip revision: 3, min. application chip revision: 0
I (137) esp_image: segment 0: paddr=00010020 vaddr=3f400020 size=407f0h (264176) map
I (241) esp_image: segment 1: paddr=00050818 vaddr=3ffb0000 size=03a44h ( 14916) load
I (247) esp_image: segment 2: paddr=00054264 vaddr=40080000 size=0bdb4h ( 48564) load
I (267) esp_image: segment 3: paddr=00060020 vaddr=400d0020 size=be1c4h (778692) map
I (549) esp_image: segment 4: paddr=0011e1ec vaddr=4008bdb4 size=0b0fch ( 45308) load
I (579) boot: Loaded app from partition at offset 0x10000
I (579) boot: Disabling RNG early entropy source...
I (591) cpu_start: Pro cpu up.
I (591) cpu_start: Starting app cpu, entry point is 0x400814ac
0x400814ac: call_start_cpu1 at C:/Users/miro/esp/esp-idf/components/esp_system/port/cpu_start.c:142

I (0) cpu_start: App cpu up.
I (607) cpu_start: Pro cpu start user code
I (608) cpu_start: cpu freq: 160000000 Hz
I (608) cpu_start: Application information:
I (612) cpu_start: Project name:     esp32-evse
I (618) cpu_start: App version:      v0.0.1-dev-24-g9689043-dirty
I (624) cpu_start: Compile time:     Feb  1 2023 22:09:39
I (630) cpu_start: ELF file SHA256:  e55420c19dd7fa16...
I (636) cpu_start: ESP-IDF:          v5.0-494-g490216a2ac
I (643) heap_init: Initializing. RAM available for dynamic allocation:
I (650) heap_init: At 3FFAE6E0 len 00001920 (6 KiB): DRAM
I (656) heap_init: At 3FFB81C8 len 00027E38 (159 KiB): DRAM
I (662) heap_init: At 3FFE0440 len 00003AE0 (14 KiB): D/IRAM
I (668) heap_init: At 3FFE4350 len 0001BCB0 (111 KiB): D/IRAM
I (675) heap_init: At 40096EB0 len 00009150 (36 KiB): IRAM
I (682) spi_flash: detected chip: generic
I (686) spi_flash: flash io: dio
I (691) cpu_start: Starting scheduler on PRO CPU.
I (0) cpu_start: Starting scheduler on APP CPU.
I (702) app_main: Running partition: factory
I (792) app_main: Partition size: total: 86846, used: 2008
E (792) board_config: Unknown config value SERIAL_1=uart
E (792) board_config: Unknown config value SERIAL_1_NAME=UART via USB
E (802) board_config: Unknown config value SERIAL_1_RXD_GPIO=3
E (802) board_config: Unknown config value SERIAL_1_TXD_GPIO=1
I (822) wifi:wifi driver task: 3ffc1b70, prio:23, stack:6656, core=0
I (822) system_api: Base MAC address is not set
I (822) system_api: read default base MAC address from EFUSE
I (852) wifi:wifi firmware version: ca38a19
I (852) wifi:wifi certification version: v7.0
I (852) wifi:config NVS flash: enabled
I (852) wifi:config nano formating: disabled
I (852) wifi:Init data frame dynamic rx buffer num: 32
I (862) wifi:Init management frame dynamic rx buffer num: 32
I (862) wifi:Init management short buffer num: 32
I (872) wifi:Init dynamic tx buffer num: 32
I (872) wifi:Init static rx buffer size: 1600
I (872) wifi:Init static rx buffer num: 10
I (882) wifi:Init dynamic rx buffer num: 32
I (882) wifi_init: rx ba win: 6
I (892) wifi_init: tcpip mbox: 32
I (892) wifi_init: udp mbox: 6
I (892) wifi_init: tcp mbox: 6
I (902) wifi_init: tcp tx win: 5744
I (902) wifi_init: tcp rx win: 5744
I (912) wifi_init: tcp mss: 1440
I (912) wifi_init: WiFi IRAM OP enabled
I (912) wifi_init: WiFi RX IRAM OP enabled
I (932) wifi: Starting STA
I (932) phy_init: phy_version 4670,719f9f6,Feb 18 2021,17:07:07
I (1042) wifi:mode : sta (78:e3:6d:64:40:74)
I (1042) wifi:enable tsf
I (1042) adc: Calibration scheme version is Line Fitting
I (1042) wifi: STA start
I (1052) gpio: GPIO[26]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (1092) energy_meter: Current zero 1647.392090 0.000000 0.000000
I (1132) energy_meter: Voltage zero 1670.236328 0.000000 0.000000
I (1132) gpio: GPIO[17]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (1142) gpio: GPIO[18]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (1152) gpio: GPIO[19]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (1162) gpio: GPIO[16]| InputEn: 0| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 
I (1182) temp_sensor: Found 1 sensors
I (1192) rest: Starting server on port: 80
E (1202) esp-tls: couldn't get hostname for :broker.hivemq.com: getaddrinfo() returns 202, addrinfo=0x0
E (1202) transport_base: Failed to open a new connection: 32769
E (1212) mqtt_client: Error transport connect
I (1212) mqtt: Disconnected
I (1222) tcp_logger: Starting server
I (1222) tcp_logger: Socket bound, port 3000
I (1232) modbus_tcp: Starting server
I (1232) modbus_tcp: Socket bound, port 502
I (1242) pilot: Set level 1
```

**Reply by dzurikmiroslav** [reply]
Are you ESP-IDF installation is right? SPIFFS example works fine https://github.com/espressif/esp-idf/tree/v5.0/examples/storage/spiffs ?

BTW latest binnary can be downloaded from project github actions  https://github.com/dzurikmiroslav/esp32-evse/actions/runs/4068904090

**Reply by vitalyxxl** [reply]
I'll try to double check everything today. Perhaps something went wrong when installing the software..... I myself am extremely interested in starting this project myself.

**Comment by Radwan-Alalili** [comment]
Hey there I hope you doing well,

I have a question, after I installed the web successfully should I compile the code and flash it to the esp32, or the web install included the code too

**Reply by dzurikmiroslav** [reply]
Hi, the web installer has compiled code inside, make esp partitions and board config as you pick. The board config can be edited even after startup via WebUI, just restart.

**Reply by Radwan-Alalili** [reply]
Awesome, 
First thank you for helping us, second can we edit the web of EVAE for example adding a new language which is Arabic, and if it is possible how can I do it.

**Reply by dzurikmiroslav** [reply]
Hi, yes there is a threat about it https://github.com/dzurikmiroslav/esp32-evse/discussions/60 
Currently working on version 1.0 with some major changes, also in WebUI. Please be patient when I publish a new `en.json` file.
